File: Sources/HedosCLI/Commands/Ls.swift
import ArgumentParser
import HedosKernel

struct Ls: AsyncParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "ls",
        abstract: "List models with fit verdict, tier, warm state, and store.")

    @OptionGroup var global: GlobalOptions

    @Flag(name: .long, help: "Rescan the machine before listing.")
    var scan = false

    @Option(name: .long, help: "Only models serving this capability (chat, image, speak, embed, ...).")
    var capability: String?

    func run() async throws {
        let kernel = Session.kernel()
        if scan { _ = try await kernel.discover() }
        var shelf = try await Session.shelf(kernel)
        if let capability {
            guard Capabilities.known.contains(capability) else {
                throw CLIError(
                    "unknown capability \"\(capability)\" — valid: "
                    + Capabilities.known.sorted().joined(separator: ", "))
            }
            let cap = Capability(rawValue: capability)
            shelf = shelf.filter { $0.capabilities.contains(cap) }
        }
        let residents = await kernel.residentModels()
        func isWarm(_ record: ModelRecord) -> Bool {
            residents.contains {
                $0.modelID == record.id || ($0.origin == .ollama && $0.name == record.name)
            }
        }
        let rows = shelf
            .sorted { $0.displayName.localizedCaseInsensitiveCompare($1.displayName) == .orderedAscending }
            .map { ModelRow($0, warm: isWarm($0)) }

        if global.json {
            try Out.json(rows)
        } else if rows.isEmpty {
            Out.line("shelf is empty — run `hedos scan`, or `hedos pull <model>` to fetch one.")
        } else {
            Out.line(ModelRow.table(rows))
        }
    }
}

struct ModelRow: Encodable {
    let id: String
    let name: String
    let store: String
    let runtime: String
    let tier: String
    let capabilities: [String]
    let fit: String?
    let warm: Bool
    let footprintMB: Int?

    init(_ record: ModelRecord, warm: Bool) {
        id = record.id
        name = record.displayName
        store = record.source.kind.rawValue
        runtime = record.runtime.id?.rawValue ?? "unresolved"
        tier = record.runtime.tier.rawValue
        capabilities = record.capabilities.map(\.rawValue)
        fit = record.fit?.verdict.label
        self.warm = warm
        footprintMB = record.footprintMB
    }

    static func table(_ rows: [ModelRow]) -> String {
        let header = ["", "NAME", "FIT", "TIER", "RUNTIME", "STORE", "CAPABILITIES"]
        var cells: [[String]] = [header]
        for row in rows {
            cells.append([
                row.warm ? "●" : "○",
                row.name,
                row.fit ?? "—",
                row.tier,
                row.runtime,
                row.store,
                row.capabilities.joined(separator: ","),
            ])
        }
        let columns = header.count
        var widths = [Int](repeating: 0, count: columns)
        for line in cells {
            for index in 0..<columns { widths[index] = max(widths[index], line[index].count) }
        }
        return cells.map { line in
            let joined = line.enumerated()
                .map { index, value in
                    index == columns - 1
                        ? value
                        : value + String(repeating: " ", count: max(0, widths[index] - value.count))
                }
                .joined(separator: "  ")
            var trimmed = Substring(joined)
            while trimmed.last == " " { trimmed = trimmed.dropLast() }
            return String(trimmed)
        }.joined(separator: "\n")
    }
}

extension FitVerdict {
    var label: String {
        switch self {
        case .runsWell: return "runs well"
        case .tightFit: return "tight"
        case .tooLarge: return "too large"
        }
    }
}

File: cli/src/commands/ls.rs
//! `hedos ls` — list the models on the shelf, with their runtime, store,
//! capabilities, and whether each is currently warm.

use clap::Args;
use kernel::records::Capability;

use crate::error::CliError;
use crate::support::output::Out;
use crate::support::session::Session;
use crate::support::shelf_table;

/// Arguments for `ls`.
#[derive(Args)]
pub struct LsArgs {
    /// Rescan the machine's model stores before listing.
    #[arg(long)]
    scan: bool,
    /// Only show models serving this capability (chat, complete, embed, …).
    #[arg(long)]
    capability: Option<String>,
}

/// Run the `ls` command.
pub async fn run(args: LsArgs, out: &Out) -> Result<(), CliError> {
    let session = Session::open()?;
    // `retain` below needs an owned, mutable `Vec`; this is the one cold CLI
    // boundary where the shared shelf snapshot is unpacked into one.
    let mut shelf = if args.scan {
        session.discover().await?;
        session.shelf().await.to_vec()
    } else {
        session.shelf_or_discover().await?.to_vec()
    };
    if let Some(name) = &args.capability {
        let capability = Capability::from(name.as_str());
        shelf.retain(|record| record.capabilities.contains(&capability));
    }

    if out.is_json() {
        out.json(&serde_json::to_value(&shelf).unwrap_or_default());
        return Ok(());
    }

    if shelf.is_empty() {
        out.line("No models on the shelf — install one with `hedos pull <ref>`.");
        return Ok(());
    }

    let warm = session.warm_set();
    let records: Vec<&_> = shelf.iter().collect();
    out.line(&shelf_table::table(&records, &warm));
    Ok(())
}
